fendermint/app/src/ipc.rs

```diff
@@ -0,0 +1,95 @@
+// Copyright 2022-2023 Protocol Labs
+// SPDX-License-Identifier: Apache-2.0, MIT
+//! IPC related execution
+
+use crate::app::{AppState, AppStoreKey};
+use crate::{App, BlockHeight};
+use bytes::Bytes;
+use fendermint_rpc::response::decode_fevm_invoke_bytes;
+use fendermint_storage::{Codec, Encode, KVReadable, KVStore, KVWritable};
+use fendermint_vm_actor_interface::{ipc, system};
+use fendermint_vm_interpreter::fvm::state::FvmStateParams;
+use fendermint_vm_interpreter::fvm::FvmMessage;
+use fendermint_vm_topdown::convert::{
+    decode_parent_finality_return, encode_get_latest_parent_finality,
+};
+use fendermint_vm_topdown::sync::ParentFinalityStateQuery;
+use fendermint_vm_topdown::{IPCParentFinality, ParentFinalityProvider};
+use fvm_ipld_blockstore::Blockstore;
+use fvm_ipld_encoding::RawBytes;
+use fvm_shared::econ::TokenAmount;
+use num_traits::Zero;
+
+/// Queries the LATEST COMMITTED parent finality from the storage
+pub struct AppParentFinalityQuery<DB, SS, S, I, P>
+where
+    SS: Blockstore + 'static,
+    S: KVStore,
+{
+    /// The app to get state
+    app: App<DB, SS, S, I, P>,
+}
+
+impl<DB, SS, S, I, P> AppParentFinalityQuery<DB, SS, S, I, P>
+where
+    S: KVStore
+        + Codec<AppState>
+        + Encode<AppStoreKey>
+        + Encode<BlockHeight>
+        + Codec<FvmStateParams>,
+    DB: KVWritable<S> + KVReadable<S> + 'static + Clone,
+    SS: Blockstore + 'static + Clone,
+    P: ParentFinalityProvider + Send + Sync,
+{
+    pub fn new(app: App<DB, SS, S, I, P>) -> Self {
+        Self { app }
+    }
+}
+
+impl<DB, SS, S, I, P> ParentFinalityStateQuery for AppParentFinalityQuery<DB, SS, S, I, P>
+where
+    S: KVStore
+        + Codec<AppState>
+        + Encode<AppStoreKey>
+        + Encode<BlockHeight>
+        + Codec<FvmStateParams>,
+    DB: KVWritable<S> + KVReadable<S> + 'static + Clone,
+    SS: Blockstore + 'static + Clone,
+    P: ParentFinalityProvider + Send + Sync,
+{
+    fn get_latest_committed_finality(&self) -> anyhow::Result<Option<IPCParentFinality>> {
+        let maybe_exec_state = self.app.new_read_only_exec_state()?;
+
+        let finality = if let Some(mut exec_state) = maybe_exec_state {
+            let msg = implicit_gateway_message(encode_get_latest_parent_finality()?);
+            tracing::debug!("query gateway parent finality message: {msg:?}");
+
+            let (apply_ret, _) = exec_state.execute_implicit(msg)?;
+
+            let data = apply_ret.msg_receipt.return_data.to_vec();
+            let decoded = decode_fevm_invoke_bytes(&Bytes::from(data))?;
```

Comment:
Actually this won't work, sorry, I was confused last time because you copied all these `decode_fevm_invoke` related methods, and didn't notice. 

It wont' work because `decode_fevm_invoke` is specifically made to parse the data inside `DeliverTx` _coming from Tendermint_ which seems to add an extra layer of Base64 encoding on top. This is explained [here](https://github.com/consensus-shipyard/fendermint/blob/main/fendermint/rpc/src/response.rs#L10-L13) and it was a result of trial and erroring and painful debugging, I am not even sure how I figured it out, by accident.

In your case, I think you do have to copy [some magic](https://github.com/consensus-shipyard/fendermint/blob/main/fendermint/rpc/src/response.rs#L53-L55), which is this one:

```rust
    fvm_ipld_encoding::from_slice::<BytesDe>(&data)
        .map(|bz| bz.0)
        .map_err(|e| anyhow!("failed to deserialize bytes returned by FEVM: {e}"))
```

This one apparently is `RawBytes` encoded as IPLD. I will never be able to remember this and I'm not sure it would be healthy to do so. I originally got this from Vyzo's testing framework I think.

Reply:
I have updated the decoding and encoding all together and they are working now